Add TransactionLayer.mkDelegationCertTx

Anviking opened the pull request:
# Issue Number

#893 

# Overview

- [x] Extended `TransactionLayer` with `mkDelegationCertTx`
- [x] Added two golden tests (replacing a temporary one committed by @KtorZ)
- [x] I've made sure the jcli instructions are correct
- [x] Added `chimericAccountFromXPub` helper

# Comments

- Related to PRs #1001 and #947 which both have review comments that are relevant here
- ℹ️ There is a commit history

<!-- Additional comments or screenshots to attach if any -->

<!-- 
Don't forget to:

 ✓ Self-review your changes to make sure nothing unexpected slipped through
 ✓ Assign yourself to the PR
 ✓ Assign one or several reviewer(s)
 ✓ Once created, link this PR to its corresponding ticket
 ✓ Assign the PR to a corresponding milestone
 ✓ Acknowledge any changes required to the Wiki
-->

Anviking commented on lib/core/src/Cardano/Wallet/Transaction.hs:
```diff
@@ -76,6 +77,21 @@ data TransactionLayer t k = TransactionLayer
         -- This expects as a first argument a mean to compute or lookup private
         -- key corresponding to a particular address.
 
+    , mkDelegationCertTx
```
I thought `mkCertificateTx` was unnecessarily vague name given that there exist multiple types of certificates. So I thought something with `delegation` or `stake` in it would be nice.

Anviking commented on lib/jormungandr/src/Cardano/Wallet/Jormungandr/Binary.hs:
```diff
@@ -569,6 +570,21 @@ constructTransactionFragment ins outs mkWit = do
+    constructGenericSignedTx FragmentDelegation
+        (putStakeCertificate pool account)
+        ins outs mkWit
+        (("\SOH" <>) . auth)
```
I didn't know what this was, but it signifies account vs multi-account signature:
```
ACCOUNT-SIG       = (%x01 SINGLE-ACNT-SIG) / (%x02 MULTI-ACNT-SIG)
```
from [format.abnf](https://github.com/input-output-hk/chain-libs/blob/master/chain-impl-mockchain/doc/format.abnf)

paweljakubas commented on lib/core/src/Cardano/Wallet/Transaction.hs:
```diff
@@ -76,6 +77,21 @@ data TransactionLayer t k = TransactionLayer
+        -- ^ Construct a transaction containing a certificate for delegating to
+        -- a stake pool.
+        --
+        -- The certificate is a combination of the node public key ('PoolId'),
```
node public key is the same as stake pool public key, right?

Anviking replied:
Hm… good point, not sure. `public key` might be a too general term too (since there are so many kinds of keys)

KtorZ replied:
> node public key is the same as stake pool public key, right?

Yes it is.

Anviking replied:
Made it "node VRF public key" at least.

paweljakubas commented on lib/jormungandr/src/Cardano/Wallet/Jormungandr/Binary.hs:
```diff
@@ -867,6 +883,7 @@ delegationFragmentId poolId accId accSig inps outs wits =
         putFragmentSpec FragmentDelegation
         putStakeDelegationTx poolId accId accSig inps outs wits
 
+
```
this adding of line is not needed I guess

paweljakubas reviewed (approved):
looks like it should work

Anviking committed:
Add TransactionLayer.mkDelegationCertTx

- Add chimericAccountFromXPub helper
- Factor shared code of mkStdTx and mkDelegationCertTx
- adjust comments

Co-authored-by: Rodney Lorrimar <rodney.lorrimar@iohk.io>

Anviking commented on lib/jormungandr/test/unit/Cardano/Wallet/Jormungandr/TransactionSpec.hs:
```diff
@@ -118,6 +113,18 @@ mkStdTxSpec = do
+    -- XPRV2HEX=e01bfb39e3e595fce0b9b19e386a82816e0cef8aa823c75a32bc0f39bcf7c14e8a03d255df0440b6d0fcf5d5199a582d1df7d858bd7556d4941ebf6223fa66d1
+    -- XPRV0=$(echo $XPRV0HEX | jcli key from-bytes --type ed25519Extended)
+    -- XPRV1=$(echo $XPRV1HEX | jcli key from-bytes --type ed25519Extended)
+    -- XPRV2=$(echo $XPRV2HEX | jcli key from-bytes --type ed25519Extended)
```
We _could_ have a bash script for generating all the goldens to files, and then `BS.readFile` from this module.

KtorZ commented on lib/core/src/Cardano/Wallet/Primitive/Types.hs:
```diff
@@ -1312,9 +1315,14 @@ newtype ProtocolMagic = ProtocolMagic Int32
 -- | Also known as a staking key, chimeric account is used in group-type address
 -- for staking purposes. It is a public key of the account address
-newtype ChimericAccount = ChimericAccount ByteString
+newtype ChimericAccount = ChimericAccount { getAccountPubKey :: ByteString }
```
**[change required]**

Rename to `getChimericAccount` or `unChimericAccount`

**Why?**

We use this pattern rather consistently for all newtypes. It allows for better identifying the underlying type at first glance instead of having to remember what `getAccountPubKey` corresponds to.

Anviking replied:
Ok. Renamed to `unChimericAccount`

 I do find it weird how we say `getOuter` to get the **inner** type. E.g: 
```haskell
newtype Earth = Earth { getEarth :: [Human] }
print $ getEarth earth -- prints a list of people
```

in contrast to how we'd name record fields:

```haskell
data Earth = Earth { people :: [Human], trees :: [Tree] }
print $ people earth
```

KtorZ commented on lib/core/src/Cardano/Wallet/Primitive/Types.hs:
```diff
@@ -1312,9 +1315,14 @@ newtype ProtocolMagic = ProtocolMagic Int32
+-- | Construct a 'ChimericAccount' by extracting the public key from the
+-- extended public key.
+chimericAccountFromXPub :: CC.XPub -> ChimericAccount
+chimericAccountFromXPub = ChimericAccount . BS.take 32 . CC.unXPub
```
:+1:

KtorZ commented on lib/core/src/Cardano/Wallet/Transaction.hs:
```diff
@@ -76,6 +77,20 @@ data TransactionLayer t k = TransactionLayer
 
+    , mkDelegationCertTx
+        :: PoolId
+        -> (k 'AddressK XPrv, Passphrase "encryption")
```
**[suggestion]**

Add a comment about the true "nature" of this `k 'AddressK XPrv` which is a chimeric account private key.

**Why?** 
It is not obvious that this refers to a chimeric account key and, somewhat confusing. Ideally, we should maybe revise the `Depth` type to have an extra path for `ChimericAccountK`. I'd suggest to that only in another PR though, a comment should be enough to dispel confusion for now.

Anviking replied:
--> 9b60c3d

Anviking replied:
Although haddock didn't like the comment as `-- ^`, so had to go with a short `-- reward account` bfc0caf0ade9759b63838b38d32b033235f80585

KtorZ commented on lib/core/src/Cardano/Wallet/Transaction.hs:
```diff
@@ -76,6 +77,20 @@ data TransactionLayer t k = TransactionLayer
+        -> (Address -> Maybe (k 'AddressK XPrv, Passphrase "encryption"))
+        -> [(TxIn, TxOut)]
+        -> [TxOut]
+        -> Either ErrMkStdTx (Tx, SealedTx)
```
**[change required]**

Use `ErrMkDelegationCertTx` instead of `ErrMkStdTx`

**Why?**

We use this pattern consistently for other functions. Plus, this function is likely to have different errors, or, errors that could be interpreted differently. As a consequence, it requires a different type to capture this.

Anviking replied:
>  Plus, this function is likely to have different errors, or, errors that could be interpreted differently.

But it doesn't? What errors are you seeing it having?

And how would be the preferred way of handling the duplication of constructors then?
```haskell
-- | Possible signing error
newtype ErrMkStdTx
    = ErrKeyNotFoundForAddress Address
    -- ^ We tried to sign a transaction with inputs that are unknown to us?
    deriving (Eq, Show)

-- | Possible signing error
newtype ErrMkDelegationCertTx
    = ErrKeyNotFoundForAddress Address
    -- ^ We tried to sign a transaction with inputs that are unknown to us?
    deriving (Eq, Show)
```

KtorZ replied:
Then, if errors are indeed the same, we should rename `ErrMkStdTx` to not suggests that it refers to `mkStdTx` but that it's a more general type of error. What about:

```hs
newtype ErrInputNotOwned 
    = ErrKeyNotFoundForAddress Address
```

Anviking replied:
> mkStdTx but that it's a more general type of error

What about simply:
```haskell
newtype ErrMkTx
    = ErrKeyNotFoundForAddress Address
```
which is used both by `mkStdTx`, and `mkDelegationCertTx`?

KtorZ replied:
As you prefer :)

Anviking replied:
Both have their qualities. I think `ErrInputNotOwned` is neat 🤷‍♂ 

But I'm defaulting to `ErrMkTx` because it is more consistent with the original `ErrMkStdTx` named after the _operation_ rather than _what could go wrong_.

KtorZ commented on lib/core/src/Cardano/Wallet/Transaction.hs:
```diff
@@ -76,6 +77,20 @@ data TransactionLayer t k = TransactionLayer
+        -- ^ Construct a transaction containing a certificate for delegating to
+        -- a stake pool.
+        --
+        -- The certificate is a combination of the node VRF public key
```
**[change required]**

This comment is now outdated. Actually, the pool id is now defined as:

`pub type PoolId = DigestOf<Blake2b256, PoolRegistration>;`

**Why?** 

Keeping comment up-to-date to avoid confusing the reader.

Anviking replied:
Also done in 9b60c3d

KtorZ commented on lib/jormungandr/src/Cardano/Wallet/Jormungandr/Transaction.hs:
```diff
@@ -70,23 +71,17 @@ newTransactionLayer
+                    (DlgFull pool)
+                    acc
+                    (first getRawKey accXPrv)
+        in mkFragment mkFrag
```
**[suggestion]**

Re-write as:

```hs
       let
            acc = chimericAccountFromXPub . getRawKey . publicKey . fst $ accXPrv
       in
            mkFragment $ MkFragmentStakeDelegation
                    (txWitnessTagFor @k)
                    (DlgFull pool)
                    acc
                    (first getRawKey accXPrv)
```

**Why?** 
The indirection is superfluous and somewhat misleading (as we expect it to be necessary). Also, this better matches the style used above which reduces cognitive load.

Anviking replied:
Done

KtorZ commented on lib/jormungandr/test/unit/Cardano/Wallet/Jormungandr/TransactionSpec.hs:
```diff
@@ -207,48 +214,69 @@ mkStdTxSpec = do
+            poolId
+            (xprv0, pwd0)
+            [(TxIn txin0 0, TxOut addr0 (Coin 42))]
+            [] -- no outputs; no change
```
**[question]**
So, the output below was generated using what kind of fee policy exactly? :thinking:

Anviking replied:
The inspiration was from our integration test setup where the fee is a constant 42.

But it seems like `jcli` doesn't require balancing transactions anymore. So it works even without passing that policy to `jcli transaction finalize`.

KtorZ replied:
Ah :|

KtorZ commented on lib/jormungandr/test/unit/Cardano/Wallet/Jormungandr/TransactionSpec.hs:
```diff
@@ -478,6 +506,27 @@ goldenTestStdTx tl keystore inps outs bytes' = it title $ do
+    let res = mkDelegationCertTx tl pool (accountXPrv, pass) keystore inps outs
+    let sealed = getSealedTx . snd <$> res
+    sealed `shouldBe` (Right $ unsafeFromHex bytes')
+    & counterexample ("poolId = " <> showHex (getPoolId pool))
```
**[question]** 
Does this `counterexample` actually works? We aren't even inside a property here, though `shouldBe` might do the trick here. How does a failure look like here?

Anviking replied:
Ehrm… Yes! c.f. https://github.com/input-output-hk/cardano-wallet/pull/1059#discussion_r350683421

KtorZ commented on lib/jormungandr/test/unit/Cardano/Wallet/Jormungandr/TransactionSpec.hs:
```diff
@@ -478,6 +506,27 @@ goldenTestStdTx tl keystore inps outs bytes' = it title $ do
+  where
+    title = "golden test mkCertificateTx: " <> show pool
+        <> show inps <> show outs
+    (&) = flip ($)
```
**[change required]**

Use [(&) from `Data.Function` ](https://hackage.haskell.org/package/base-4.12.0.0/docs/Data-Function.html#v:-38-) 

**Why?**
Avoid superfluous definition from a function defined the `base` package.

Anviking replied:
Done

KtorZ commented on lib/jormungandr/test/unit/Cardano/Wallet/Jormungandr/TransactionSpec.hs:
```diff
@@ -478,6 +506,27 @@ goldenTestStdTx tl keystore inps outs bytes' = it title $ do
+goldenTestDelegationCertTx tl keystore pool (accountXPrv, pass) inps outs bytes' = it title $ do
+    let res = mkDelegationCertTx tl pool (accountXPrv, pass) keystore inps outs
+    let sealed = getSealedTx . snd <$> res
+    sealed `shouldBe` (Right $ unsafeFromHex bytes')
```
**[suggestion]**

Compare hex-encoded bytestrings instead of decoded bytestrings

**Why?**

This makes the counter-example output automatically more readable via `shouldBe` (for what it's worth). Also, this is more consistent with the other golden test:

```hs
goldenTestStdTx tl keystore inps outs bytes' = it title $ do
    let tx = mkStdTx tl keystore inps outs
    let bytes = hex . getSealedTx . snd <$> tx
    bytes `shouldBe` Right bytes'
```

Anviking replied:
From spending much time looking at these failures, I found the hex-encoded version unusable because it doesn't highlight where the changes are within the strings.

Looking at the raw byte string gives me a quick indication of how much/where the differences are in the binary. If I needed to have a closer look, I'd write them to files and compare with [HexFiend](http://ridiculousfish.com/hexfiend/).

It feels like we should be able to have the same highlighting when hex-encoded, but I'm unable to find any trace on google.

<img width="1440" alt="Skärmavbild 2019-11-26 kl  12 02 32" src="https://user-images.githubusercontent.com/304423/69624598-4551d480-1045-11ea-9d5f-8a534a3f578e.png">
<img width="1440" alt="Skärmavbild 2019-11-26 kl  12 06 57" src="https://user-images.githubusercontent.com/304423/69624600-471b9800-1045-11ea-9e50-062e5ef129aa.png">

KtorZ replied:
Indeed. That's weird. I've tried using some alternatives by comparing list instead (e.g. `BS.unpack bytes' shouldBe BS.unpack bytes`) but the output is kinda unsatisfactory.

KtorZ reviewed (approved):
LGTM, one naming issue to fix however regarding `MkStdTx` which is now a misleading.

Anviking committed:
Rename ErrMkStdTx to ErrMkTx
